### PR TITLE
Add SMask to GraphicalState consuming a transparency Group.

### PR DESCRIPTION
**Pull request by doctormo**
Fixes https://github.com/jpakkane/capypdf/issues/50

### CODE REVIEW

File: src/document.cpp [hunk 1]
```diff
@@ -1666,6 +1669,19 @@ rvoe<CapyPDF_TransparencyGroupId> PdfDocument::add_transparency_group(PdfDrawCon
     return CapyPDF_TransparencyGroupId{(int32_t)transparency_groups.size() - 1};
 }
 
+rvoe<CapyPDF_SoftMaskId> PdfDocument::add_soft_mask(CapyPDF_TransparencyGroupId tgid) {
+    auto id = add_object(FullPDFObject{std::format(R"(<<
+  /Type /Mask
+  /S /Luminosity
+  /G {} 0 R
+>>
```
Comment (jpakkane) [inline src/document.cpp:1677]: The basic outline of the change is good, but this hardcodes the mask type, which is not ok, we need to support all formats. This requires a new top level object. Thus the implementation needs to do roughly:

- add a new struct to `pdfcommon.hpp` called `SoftMask` that has entries S and G (and maybe comment to add BC and TR later)
- the constructor function should take both of those as arguments
- there needs to be a new enum for choosing between luminosity and alpha
- eventually the API should have functions for setting the optional entries, like `set_BC` and `set_TR`, but they are not needed for this, they can be added once someone actually needs them

**Comment by doctormo** [comment]
OK lets give this version a go.

File: src/document.cpp [hunk 2]
```diff
@@ -1669,6 +1673,22 @@ rvoe<CapyPDF_TransparencyGroupId> PdfDocument::add_transparency_group(PdfDrawCon
     return CapyPDF_TransparencyGroupId{(int32_t)transparency_groups.size() - 1};
 }
 
+rvoe<CapyPDF_SoftMaskId> PdfDocument::add_soft_mask(const SoftMask &sm) {
+    std::string subtype = sm.S == CAPY_SOFT_MASK_ALPHA ? "Alpha" : "Luminosity";
```
Comment (jpakkane) [inline src/document.cpp:1677]: This creates a temporary copy of the string that is then immediately destroyed. The ternary should be directly in the format call below.

Comment (doctormo) [reply src/document.cpp]: Fixed.

File: src/document.cpp (hunk 1 above)
Comment (doctormo) [reply src/document.cpp:1681]: Thanks for the notes, this helped reformat the patch.

File: src/document.cpp [hunk 3]
```diff
@@ -1669,6 +1673,20 @@ rvoe<CapyPDF_TransparencyGroupId> PdfDocument::add_transparency_group(PdfDrawCon
...
 }
 
+rvoe<CapyPDF_SoftMaskId> PdfDocument::add_soft_mask(const SoftMask &sm) {
+    auto id = add_object(FullPDFObject{std::format(R"(<<
+  /Type /Mask
+  /S /{}
+  /G {} 0 R
+>>
+)",
+                                                   CAPY_SOFT_MASK_ALPHA ? "Alpha" : "Luminosity",
```
[the remark this answers is not in the document]
Comment (doctormo) [reply src/document.cpp]: Woops.

**Comment by jpakkane** [comment]
Implementation-wise looks good, but it would be nice to get a test for this if one can be created with reasonable effort.

**Comment by doctormo** [comment]
I've added all the python API support and a python test. Yikes the API is really verbose when you're doing gradients.

**Comment by doctormo** [comment]
Fixed the test file.

Oh and here's a meme I found :wink: : [image](https://github.com/user-attachments/assets/79872e94-403b-4f72-a4f4-c090df47880f)

**Comment by jpakkane** [comment]
> Yikes the API is really verbose when you're doing gradients.

This is a problem I struggle with in general. PDF itself is quite verbose and when you combine that with ABI stability requirements, usage ergonomics tend to suffer. I have tried to simplify the API whenever possible. Ideas on how to make it less cumbersome are welcome.

Though in practice I'd expect most projects to write some simple wrapper code to generate the things they need and then forget about it.

**Comment by doctormo** [comment]
Pretty much what we have in Inkscape, I don't expect it to be a problem. Some things might be possible to tighten, like building a Color object in python, you could have `capypdf.Color.from_rgb(...)` instead of having two lines.

It depends how dissimilar you want the python to be from your C.

**Pull request merged**